pkg/controller/volume/attachdetach/util/util_test.go

```diff
@@ -0,0 +1,369 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubetypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/informers"
+	csitrans "k8s.io/csi-translation-lib"
+	fakeframework "k8s.io/kubernetes/pkg/scheduler/framework/fake"
+	"k8s.io/kubernetes/pkg/volume/csimigration"
+	"k8s.io/kubernetes/pkg/volume/fc"
+
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/kubernetes/fake"
+	utiltesting "k8s.io/client-go/util/testing"
+	"k8s.io/kubernetes/pkg/volume"
+	volumetest "k8s.io/kubernetes/pkg/volume/testing"
+)
+
+const (
+	testHostName = "test-hostname"
+	socketPath   = "/var/run/kmsplugin"
+	volumeName   = "volume-name"
+	volumeName2  = "volume-name-2"
+	testNodeName = "test-node-name"
+)
+
+var (
+	emptyVolume           = v1.Volume{}
+	emptyPersistentVolume = v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{},
+	}
+	dirOrCreate = v1.HostPathType(v1.HostPathDirectoryOrCreate)
+	nodeName    = kubetypes.NodeName(testNodeName)
+	hostPath    = &v1.HostPathVolumeSource{
+		Path: socketPath,
+		Type: &dirOrCreate,
+	}
+	objectReference  = v1.ObjectReference{Namespace: "claimrefns", Name: "test-pvc"}
+	objectReference2 = v1.ObjectReference{Namespace: "claimrefns", Name: "test-pv"}
+
+	volumeExample = v1.Volume{
+		Name: volumeName,
+		VolumeSource: v1.VolumeSource{
+			HostPath: hostPath,
+		},
+	}
+	volumeExampleWithCSI = v1.Volume{
+		Name: volumeName,
+		VolumeSource: v1.VolumeSource{
+			PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+				ClaimName: "test-pvc",
+				ReadOnly:  false,
+			},
+		},
+	}
+	pvExampleWithCSI = v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: volumeName,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				CSI: &v1.CSIPersistentVolumeSource{
+					Driver:           "pd.csi.storage.gke.io",
+					VolumeHandle:     "projects/UNSPECIFIED/zones/UNSPECIFIED/disks/test-disk",
+					FSType:           "ext4",
+					ReadOnly:         false,
+					VolumeAttributes: map[string]string{"partition": ""},
+				},
+			},
+			ClaimRef: &objectReference,
+		},
+	}
+)
+
+type vaTest struct {
+	desc                 string
+	seupNodeName         string
+	createNodeName       kubetypes.NodeName
+	podVolume            v1.Volume
+	pod                  *v1.Pod
+	wantVolume           v1.Volume
+	wantPersistentVolume v1.PersistentVolume
+	wantErrorMessage     string
+}
+
+func Test_CreateVolumeSpec(t *testing.T) {
+	for _, test := range []vaTest{
+		{
+			desc:           "inline volume type that does not support csi migration",
+			seupNodeName:   testNodeName,
+			createNodeName: nodeName,
+			podVolume:      volumeExample,
```

Comment:
it looks like `podVolume` is already included in `pod`. Can we get rid of `podVolume`?